File: src/main/java/com/github/tianchenghang/ai/tools/CodeQuestionTool.java
package com.github.tianchenghang.ai.tools;

import static org.slf4j.LoggerFactory.getLogger;

import dev.langchain4j.agent.tool.P;
import dev.langchain4j.agent.tool.Tool;
import java.io.IOException;
import java.net.URLEncoder;
import java.nio.charset.StandardCharsets;
import java.util.ArrayList;
import org.jsoup.Jsoup;
import org.jsoup.nodes.Document;
import org.slf4j.Logger;

public class CodeQuestionTool {
  private static final Logger log = getLogger(CodeQuestionTool.class);

  @Tool(
      name = "CodeQuestionTool",
      value =
          """
      Find relevant code questions based on a keyword.
      Use this tool when the user asks for code questions.
      The input should be a clear search keyword.
      """)
  public String searchCodeQuestions(@P(value = "The keyword to search") String keyword) {
    var questions = new ArrayList<String>();
    var encodedKeyword = URLEncoder.encode(keyword, StandardCharsets.UTF_8);
    var url = "https://leetcode.cn/search/?q=" + encodedKeyword;
    log.info("Found relevant code questions from leetcode.cn for keyword: {}", keyword);
    Document doc;
    try {
      doc = Jsoup.connect(url).userAgent("Mozilla/5.0").timeout(5000).get();
    } catch (IOException err) {
      log.error("Error connecting to leetcode.cn", err);
      return err.getMessage();
    }
    var elems = doc.select("a");
    elems.forEach(el -> questions.add(el.text().trim()));
    return String.join("\n", questions);
  }
}

File: ai-js/src/ai/tools/code-question-tool.ts
import { StructuredTool } from '@langchain/core/tools';
import { Injectable, Logger } from '@nestjs/common';
import cheerio from 'cheerio';
import { z } from 'zod';

const CodeQuestionInputSchema = z.object({
  keyword: z.string().describe('The keyword to search'),
});

type CodeQuestionInput = z.infer<typeof CodeQuestionInputSchema>;

@Injectable()
export class CodeQuestionTool extends StructuredTool<
  typeof CodeQuestionInputSchema
> {
  private readonly logger = new Logger(CodeQuestionTool.name);

  name = 'CodeQuestionTool';

  description = `Find relevant code questions based on a keyword.
Use this tool when the user asks for code questions.
The input should be a clear search keyword.`;

  schema = CodeQuestionInputSchema;

  async _call(input: CodeQuestionInput): Promise<string> {
    const { keyword } = input;
    const questions: string[] = [];
    const encodedKeyword = encodeURIComponent(keyword);
    const url = `https://leetcode.cn/search/?q=${encodedKeyword}`;
    try {
      const controller = new AbortController();
      setTimeout(() => {
        controller.abort();
      }, 5000);
      const response = await fetch(url, {
        headers: {
          'User-Agent': 'Mozilla/5.0',
        },
        signal: controller.signal,
      });
      if (!response.ok) {
        throw new Error(`HTTP ${response.status}`);
      }
      const html = await response.text();
      const $ = cheerio.load(html);
      $('a').each((_, el) => {
        const text = $(el).text().trim();
        if (text) {
          questions.push(text);
        }
      });
      this.logger.log(
        `Found relevant ${questions.length} code questions from leetcode.cn for keyword: ${keyword}`,
      );
      return questions.join('\n') ?? 'No questions found';
    } catch (err) {
      this.logger.error('Code question search failed:', err);
      return `Code question search failed: ${err.message}`;
    }
  }
}
